nilearn/_utils/niimg.py

```diff
@@ -0,0 +1,157 @@
+"""
+Neuroimaging file input and output.
+"""
+# Author: Gael Varoquaux, Alexandre Abraham, Philippe Gervais
+# License: simplified BSD
+
+import copy
+import gc
+import collections
+from six import string_types
+
+import numpy as np
+import nibabel
+
+from .numpy_conversions import as_ndarray
+
+
+def _safe_get_data(img):
+    """ Get the data in the image without having a side effect on the
+        Nifti1Image object
+    """
+    if hasattr(img, '_data_cache') and img._data_cache is None:
+        # Copy locally the Nift1Image object to avoid the side effect of data
+        # loading
+        img = copy.deepcopy(img)
+    # typically the line below can double memory usage
+    # that's why we invoke a forced call to the garbage collector
+    gc.collect()
+    return img.get_data()
+
+
+def is_img(obj):
+    """ Check for get_data and get_affine method in an object
+
+    Parameters
+    ----------
+    obj: any object
+        Tested object
+
+    Returns
+    -------
+    is_img: boolean
+        True if get_data and get_affine methods are present and callable,
+        False otherwise.
+    """
+
+    # We use a try/except here because this is the way hasattr works
+    try:
+        get_data = getattr(obj, "get_data")
+        get_affine = getattr(obj, "get_affine")
+        return callable(get_data) and callable(get_affine)
+    except AttributeError:
+        return False
+
+
+# XXX add dtype option for masks
+def load_img(img):
+    """Load a niimg and check if it has required methods
+    """
+    if isinstance(img, string_types):
+        # data is a filename, we load it
+        img = nibabel.load(img)
+    elif not is_img(img):
+        raise TypeError("Data given cannot be converted to a nifti"
+                        " image: this object -'%s'- does not expose"
+                        " get_data or get_affine methods"
+                        % short_repr(img))
+    return img
+
+
+def new_img_like(ref_img, data, affine, copy_header=False):
+    """Create a new image of the same class as the reference image
+
+    Parameters
+    ----------
+    ref_img: Niimg-like image
+        Reference image. The new image will be of the same type.
+
+    data: numpy array
+        Data to be stored in the image
+
+    affine: 4x4 numpy array
+        Transformation matrix
+
+    copy_header: boolean, optional
+        Indicated if the header of the reference image should be used to
+        create the new image
+
+    Returns
+    -------
+
+    new_img: nibabel.SpatialImage
+        An image which has the same type as the reference image.
+    """
+    # XXX If the niimg is a list of 3D images, we don't need to load them all
+    # But this is a bit ugly
+    if hasattr(ref_img, "__iter__") and not isinstance(ref_img, string_types):
+        ref_img = next(ref_img)
+    ref_img = load_img(ref_img)
+    header = ref_img.get_header() if copy_header else None
```

Comment:
> I'd rather not do that in this PR. It has already gone far beyond what I wanted
> to do and I guess that the question of which header to keep / update / reset
> will be subject to debate. 

Well, the problem is that not resetting some values will give you bugs
that are very hard to debug with the softwares that respect these values.

I suggest that you put 'cal_min', 'cal_max' to zero (which means None in
Nifti terms), or better, to the min and max.

In general, I might even do the following:
{'scl_slope':0, 'scl_inter':0, 'glmax':0, 'glmin':0, 'cal_max':0,
'cal_min':0}

Once again: these things are really tricky, and if you don't do it right,
you'll end up with bugs. I am actually very weary of the nifti header.